src/transformers/trainer.py

```diff
@@ -0,0 +1,490 @@
+        checkpoints_sorted = [checkpoint[1] for checkpoint in checkpoints_sorted]
+        return checkpoints_sorted
+
+    def _rotate_checkpoints(self, use_mtime=False) -> None:
+        if not self.args.save_total_limit:
+            return
+        if self.args.save_total_limit <= 0:
+            return
+
+        # Check if we should delete older checkpoint(s)
+        checkpoints_sorted = self._sorted_checkpoints(use_mtime=use_mtime)
+        if len(checkpoints_sorted) <= self.args.save_total_limit:
+            return
+
+        number_of_checkpoints_to_delete = max(0, len(checkpoints_sorted) - self.args.save_total_limit)
+        checkpoints_to_be_deleted = checkpoints_sorted[:number_of_checkpoints_to_delete]
+        for checkpoint in checkpoints_to_be_deleted:
+            logger.info("Deleting older checkpoint [{}] due to args.save_total_limit".format(checkpoint))
+            shutil.rmtree(checkpoint)
+
+    def evaluate(self, eval_dataset: Optional[Dataset] = None, return_loss_only=False,) -> Dict[str, float]:
+        """
+        Run evaluation and return results.
+
+        The calling script will be responsible for computing metrics, as they are
+        task-dependent.
+
+        Args:
+            eval_dataset: (Optional) Pass a dataset if you wish to override
+            the one on the instance.
+        Returns:
+            A dict containing:
+                - the eval loss
+                - the potential metrics computed from the predictions
+        """
+        eval_dataloader = self.get_eval_dataloader(eval_dataset)
+
+        # multi-gpu eval
+        if self.args.n_gpu > 1 and not isinstance(self.model, torch.nn.DataParallel):
+            model = torch.nn.DataParallel(self.model)
+        else:
+            model = self.model
+        model.to(self.args.device)
+
+        # Eval!
+        logger.info("***** Running evaluation *****")
+        logger.info("  Num examples = %d", len(eval_dataloader.dataset))
+        logger.info("  Batch size = %d", self.args.eval_batch_size)
+        eval_loss = 0.0
+        nb_eval_steps = 0
+        preds: np.ndarray = None
+        label_ids: np.ndarray = None
+        model.eval()
+
+        for inputs in tqdm(eval_dataloader, desc="Evaluating"):
+            for k, v in inputs.items():
+                inputs[k] = v.to(self.args.device)
+
+            with torch.no_grad():
+                outputs = model(**inputs)
+                step_eval_loss, logits = outputs[:2]
+
+                eval_loss += step_eval_loss.mean().item()
+            nb_eval_steps += 1
+            if not return_loss_only:
+                if preds is None:
+                    preds = logits.detach().cpu().numpy()
+                    label_ids = inputs["labels"].detach().cpu().numpy()
+                else:
+                    preds = np.append(preds, logits.detach().cpu().numpy(), axis=0)
+                    label_ids = np.append(label_ids, inputs["labels"].detach().cpu().numpy(), axis=0)
```

Comment:
I would move this to an `self._evaluation_step()` that could be overridden by inherited Trainers (for instance to be able to call `model.generate()` to train/evaluate on summarization).

Reply:
Yes, good point